pkg/consensus/consensus.go

```diff
@@ -1,34 +1,367 @@
+		dbft.WithGetConsensusAddress(srv.getConsensusAddress),
+
+		dbft.WithNewConsensusPayload(func() payload.ConsensusPayload { return new(Payload) }),
+		dbft.WithNewPrepareRequest(func() payload.PrepareRequest { return new(prepareRequest) }),
+		dbft.WithNewPrepareResponse(func() payload.PrepareResponse { return new(prepareResponse) }),
+		dbft.WithNewChangeView(func() payload.ChangeView { return new(changeView) }),
+		dbft.WithNewCommit(func() payload.Commit { return new(commit) }),
+	)
+
+	if srv.dbft == nil {
+		return nil, errors.New("can't initialize dBFT")
+	}
+
+	return srv, nil
+}
+
+var (
+	_ block.Transaction = (*transaction.Transaction)(nil)
+	_ block.Block       = (*neoBlock)(nil)
+)
+
+func (s *service) Start(ctx context.Context) {
+	s.dbft.Start(ctx)
+
+	go s.eventLoop(ctx)
+}
+
+func (s *service) eventLoop(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			s.log.Debugf("context expired")
+		case hv := <-s.dbft.Timer.C():
+			s.log.Debugf("timer fired (%d,%d)", hv.Height, hv.View)
+			s.dbft.OnTimeout(ctx, hv)
+		case msg := <-s.messages:
+			s.log.Debugf("received message from %d", msg.validatorIndex)
+			s.dbft.OnReceive(ctx, &msg)
+		}
+	}
+}
+
+func getKeyPair(cfg *config.WalletConfig) (crypto.PrivateKey, crypto.PublicKey) {
+	acc, err := wallet.DecryptAccount(cfg.Path, cfg.Password)
+	if err != nil {
+		return nil, nil
 	}
+
+	key := acc.PrivateKey()
+	if key == nil {
+		return nil, nil
+	}
+
+	return &privateKey{PrivateKey: key}, &publicKey{PublicKey: key.PublicKey()}
 }
 
 // OnPayload handles Payload receive.
-func (s *service) OnPayload(p *Payload) {
-	s.cache.Add(p)
+func (s *service) OnPayload(cp *Payload) {
+	if s.cache.Has(cp.Hash()) {
+		return
+	}
+
+	s.Config.Broadcast(cp)
+	s.cache.Add(cp)
+
+	// we use switch here because other payloads could be possibly added in future
+	switch cp.Type() {
+	case payload.PrepareRequestType:
+		s.txx.Add(&cp.GetPrepareRequest().(*prepareRequest).minerTx)
+	}
+
+	s.messages <- *cp
 }
 
 // GetPayload returns payload stored in cache.
 func (s *service) GetPayload(h util.Uint256) *Payload {
-	return s.cache.Get(h)
+	p := s.cache.Get(h)
+	if p == nil {
+		return (*Payload)(nil)
+	}
+
+	cp := *p.(*Payload)
+
+	return &cp
+}
+
+func (s *service) broadcast(ctx context.Context, p payload.ConsensusPayload) {
+	switch p.Type() {
+	case payload.PrepareRequestType:
+		pr := p.GetPrepareRequest().(*prepareRequest)
+		pr.minerTx = *s.txx.Get(pr.transactionHashes[0]).(*transaction.Transaction)
+	}
+
+	s.cache.Add(p)
+	s.Config.Broadcast(p.(*Payload))
+}
+
+func (s *service) getTx(h util.Uint256) block.Transaction {
+	if tx := s.txx.Get(h); tx != nil {
+		return tx.(*transaction.Transaction)
+	}
+
+	tx, _, _ := s.Config.Chain.GetTransaction(h)
+
+	return tx
+}
+
+func (s *service) verifyBlock(b block.Block) bool {
+	coreb := &b.(*neoBlock).Block
+	for _, tx := range coreb.Transactions {
+		if err := s.Chain.VerifyTx(tx, coreb); err != nil {
+			return false
+		}
+	}
```

Comment:
1. Validity of all other fields is guaranteed by the library, because the block is created by ourselves, only transactions come from external world.
2. Do you mean just checking tx presence in memory pool? What if a transaction was received but not yet verified?

Reply:
Looks like I'm missing something here. If you're creating this block, you're probably using known verified transactions from the mempool, aren't you? If it's not yet verified it just doesn't go into the block.